### ISSUE TITLE
Auto-translate all Box REST API objects

### ISSUE THREAD

**Issue by jmoldow** [box/box-python-sdk#140]
Box REST API endpoints usually return JSON. Most of these are either objects (with a `type` field and an `id` field), "pseudo-objects" (something with a `type`, but not an `id`, e.g. an `event`), or collections (e.g. the responses from `/folder/:id:/items` or `/events`). Furthermore, they will all usually have references to other objects (e.g. the `parent` of an item, the `source` of an `event`, or the `entries` of an `/events` response).

By default, all of these will be returned from the SDK as JSON `dict` objects. The only places were we do translations into the smart instances of `BaseObject` subclasses is in locations where we explicitly do the translation ourselves:
- https://github.com/box/box-python-sdk/blob/3ba6cf1fe50f2f9b3c273d96375dfdc98ea87d3c/boxsdk/object/group_membership.py#L73
- https://github.com/box/box-python-sdk/blob/3ba6cf1fe50f2f9b3c273d96375dfdc98ea87d3c/boxsdk/object/search.py#L232
- https://github.com/box/box-python-sdk/blob/3ba6cf1fe50f2f9b3c273d96375dfdc98ea87d3c/boxsdk/object/base_object.py#L240
- https://github.com/box/box-python-sdk/blob/3ba6cf1fe50f2f9b3c273d96375dfdc98ea87d3c/boxsdk/client/client.py#L292
- https://github.com/box/box-python-sdk/blob/3ba6cf1fe50f2f9b3c273d96375dfdc98ea87d3c/boxsdk/object/folder.py#L155

It'd be great if this translation could happen automatically, for arbitrarily nested objects, and remove most of these manual translations.

Since pseudo-objects and collections are not REST objects, they cannot be instances of `BaseObject`. So this functionality cannot be written directly into `BaseObject`. Instead, we can introduce a new mixin with this functionality (perhaps called `AutomaticTranslatorMixin`). We could implement it in one of two ways:
- During `__init__`, traverse the entire JSON structure and translate everything.
- Override `__getattr__` and `__getitem__` to do translation on demand.

A new class, `BaseAPIJSONObject`, would inherit from `AutomaticTranslatorMixin`. And `BaseObject` would inherit from `BaseAPIJSONObject`.

Pseudo-objects and collections would need to be translated into `BaseAPIJSONObject` instances, not `BaseObject` instances. However, 
- We would like to translate psudo-objects into more specific subclass instances (e.g. `Event`, see #15), in the same way that we currently do translation for `BaseObject` subclasses.
- We have a backwards compatibility constraint: existing endpoints that returns pseudo-objects and collections return `dict` instances, and our users are certainly using them as `Mapping`s or `MutableMapping`s.

So the translation registration logic would move from `BaseObjectMeta` to `BaseAPIJSONObjectMeta`, in order to accommodate translation for both objects and pseudo-objects. And for the backwards compatibility issue, we have this class hierarchy:

``` python
class AutomaticTranslatorMixin(object): pass

class BaseAPIJSONObjectMeta(type):
    # Renamed from BaseObjectMeta
    pass

class BaseAPIJSONObject(AutomaticTranslatorMixin, metaclass=BaseAPIJSONObjectMeta):
    # Move some, but not all, of the logic from BaseObject into here.
    # There would not directly be any instances of this, just instances of subclasses.
    # NOTE: Not a Mapping.
    pass

class BaseObject(BaseAPIJSONObject, BaseEndpoint):
    # Only keep the logic related to being an addressable REST object.
    # NOTE: Under this proposal, this stays as it is now:
    # not a Mapping, just a normal object that happens to have a __getitem__ method.
    pass

class APIJSONObject(BaseAPIJSONObject, Mapping):
    # Pseudo-objects and collections would be instances of this.
    # Pseudo-object classes for translation would be subclasses of this.
    # NOTE: Possible mapping baseclasses, in order from least to most backwards compatible:
    # Mapping < MutableMapping < dict
```

**Comment by Jeff-Meadows** [comment]
Similar to what I was thinking. Separate the REST interaction from the attribute access/translation behaviors into two base classes. `Event` only needs the latter.

**Comment by jmoldow** [comment]
`BaseObject` definitely shouldn't be a `dict` or a `MutableMapping`. I also realized that we shouldn't make it a `Mapping`, because they have different notions of the method `get`. So we shouldn't go with the first of the suggested class hierarchies. I'll remove it from the text of the issue.

**Comment by mattwiller** [comment]
This functionality was released with v2.0.0
